suitesparse_bindings/sprs_suitesparse_umfpack/src/lib.rs

```diff
@@ -0,0 +1,380 @@
+        struct $Symbolic(*mut c_void);
+
+        impl Drop for $Symbolic {
+            fn drop(&mut self) {
+                unsafe {$free_symbolic(&mut self.0 as *mut *mut c_void);}
+            }
+        }
+
+        /// Wrapper of raw handle to guarantee proper drop procedure
+        struct $Numeric(*mut c_void);
+
+        impl Drop for $Numeric {
+            fn drop(&mut self) {
+                unsafe {$free_numeric(&mut self.0 as *mut *mut c_void);}
+            }
+        }
+
+        /// Partial interface to `SuiteSparse`'s UMFPACK solver package.
+        /// Provides LU factorization of A and solution of Ax=b using stored factorization.
+        pub struct $Context {
+            /// `A` matrix of system Ax=b
+            _a: CsMatI<f64, $int>,
+            /// Opaque raw handle to symbolic factorization
+            #[allow(dead_code)]  // We don't use this at the moment, but could extend the bindings to extract state
+            symbolic: $Symbolic,
+            /// Opaque raw handle to numeric factorization
+            numeric: $Numeric,
+            /// Number of rows in `A`
+            nrow: usize,
+            /// Number of cols in `A`
+            ncol: usize,
+            /// Number of nonzeroes in `A`
+            _nnz: usize,
+        }
+
+        impl $Context {
+
+            /// Build a new stored factorization of matrix `A` into LU components
+            /// with a stored C handle to an efficient (but opaque) direct solver.
+            ///
+            /// This factorization can be done for either square or rectangular `A` matrix,
+            /// but can only be solved directly if `A` is square.
+            pub fn new<N>(a: CsMatI<N, $int>) -> Self
+            where N: Default + Clone + Into<f64>,
+            {
+                // UMFPACK methods require CSC format and f64 data
+                let a: CsMatI<f64, $int> = a.to_other_types().into_csc();
+
+                // Get shape info
+                let nrow_ = a.rows();
+                let ncol_ = a.cols();
+                let nnz = a.nnz();
+
+                // Get C-compatible raw pointers to column pointers, indices, and values
+                let ap = a.indptr().to_proper().as_ptr();
+                let ai = a.indices().as_ptr();
+                let ax = a.data().as_ptr();
+
+                // Do symbolic factorization
+                let symbolic_inner = &mut (null_mut() as *mut c_void) as *mut *mut c_void;
+                unsafe {
+                    $symbolic(
+                        nrow_ as $int,
+                        ncol_ as $int,
+                        ap,
+                        ai,
+                        ax,
+                        symbolic_inner,
+                        null() as *const f64,  // Default settings
+                        null_mut() as *mut c_void  // Ignore info
+                    );
+                };
+                let symbolic_ = unsafe{$Symbolic(*symbolic_inner)};
```

Comment:
Drop trailing underscore

Reply:
So, tying this in with the other thought about clippy - I had to add underscores to these because clippy was giving me lints on these at the initialization of $Context. It gets tricked by repeated words that it thinks are repeated fields, like

Struct {
    x: x  // Here, it thinks the second `x` is a repeated field instead of a value with the same name
}

Reply:
The tip says to have just `dx`, giving struct declarations like
```rust
                $NumericComponents {
                    l,
                    u,
                    p,
                    q,
                    rs,
                    dx
                }
```

Reply:
That's because I made an example to show that error - the actual definitions all have it fixed. There is an alternative to the trailing underscores, which is to go to every single location where it causes a problem and tell clippy to ignore it, but this seems cleaner

Reply:
Ok, I figured out what was going on here and was able to get rid of all the trailing underscores